src/CommunityToolkit.Maui/Views/Popup/PopupPage.shared.cs

```diff
@@ -0,0 +1,185 @@
+
+sealed partial class PopupPage<T>(Popup<T> popup, IPopupOptions popupOptions) : PopupPage(popup, popupOptions)
+{
+	public PopupPage(View view, IPopupOptions popupOptions)
+		: this(view as Popup<T> ?? CreatePopupFromView<Popup<T>>(view), popupOptions)
+	{
+	}
+
+	public Task Close(PopupResult<T> result, CancellationToken token = default) => base.Close(result, token);
+}
+
+partial class PopupPage : ContentPage
+{
+	readonly Popup popup;
+	readonly IPopupOptions popupOptions;
+	readonly Command tapOutsideOfPopupCommand;
+	readonly WeakEventManager popupClosedEventManager = new();
+
+	public PopupPage(View view, IPopupOptions popupOptions)
+		: this(view as Popup ?? CreatePopupFromView<Popup>(view), popupOptions)
+	{
+		ArgumentNullException.ThrowIfNull(view);
+	}
+
+	public PopupPage(Popup popup, IPopupOptions popupOptions)
+	{
+		ArgumentNullException.ThrowIfNull(popup);
+		ArgumentNullException.ThrowIfNull(popupOptions);
+
+		this.popup = popup;
+		this.popupOptions = popupOptions;
+
+		// Only set the content if parent constructor hasn't set the content already; don't override content if it already exists
+		base.Content ??= new PopupPageLayout(popup, popupOptions);
+
+		tapOutsideOfPopupCommand = new Command(async () =>
+		{
+			popupOptions.OnTappingOutsideOfPopup?.Invoke();
+			await Close(new PopupResult(true));
+		}, () => popupOptions.CanBeDismissedByTappingOutsideOfPopup);
+
+		Content.GestureRecognizers.Add(new TapGestureRecognizer { Command = tapOutsideOfPopupCommand });
+
+		if (popupOptions is BindableObject bindablePopupOptions)
+		{
+			bindablePopupOptions.PropertyChanged += HandlePopupPropertyChanged;
+		}
+
+		this.SetBinding(BindingContextProperty, static (Popup x) => x.BindingContext, source: popup, mode: BindingMode.OneWay);
+		this.SetBinding(BackgroundColorProperty, static (IPopupOptions options) => options.PageOverlayColor, source: popupOptions, mode: BindingMode.OneWay);
+
+		Shell.SetPresentationMode(this, PresentationMode.ModalNotAnimated);
+		On<iOS>().SetModalPresentationStyle(UIModalPresentationStyle.OverFullScreen);
+	}
+
+	public event EventHandler<IPopupResult> PopupClosed
+	{
+		add => popupClosedEventManager.AddEventHandler(value);
+		remove => popupClosedEventManager.RemoveEventHandler(value);
+	}
+
+	// Prevent Content from being set by external class
+	// Casts `PopupPage.Content` to return typeof(PopupPageLayout)
+	internal new PopupPageLayout Content => (PopupPageLayout)base.Content;
+
+	public async Task Close(PopupResult result, CancellationToken token = default)
+	{
+		token.ThrowIfCancellationRequested();
+
+		var popupPageToClose = Navigation.ModalStack.OfType<PopupPage>().LastOrDefault(popupPage => popupPage.Content == Content);
+
+		if (popupPageToClose is null)
+		{
+			throw new InvalidOperationException($"Unable to close popup: could not locate {nameof(PopupPage)}. If using a custom implementation of {nameof(Popup)}, override the {nameof(Close)} method");
```

Comment:
not sure if should throw for here, I can't see good reason to do an exception if the popup can't be closed and not given a way for devs to leave here. In other words, let's say that we have 3 popups and for some reason I can't close the last one, then my app will crash (if the exception isn't handled) or will be blocked on those popups forever.

Maybe we can throw and add a method `CloseAllPopups` that will close all popups opened popups?

Reply:
The Exception is valid. 

When the developer calls `Close()`, we have two options:
1. Close the current Popup
2. Throw an exception telling them why we were unable to close the Popup

Reply:
Yeap, but in a production code, the app will be blocked forever on the popup. So I would say that we should provide a `panic mode`, a method that will close all popups.